device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,166 @@
+                connectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+    }
+
+    public void open() throws IOException {
+        synchronized (lock) {
+            if(connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+        Pair<IotHubConnectionStatusChangeCallback, Object> suppliedCallbackPair = this.suppliedConnectionStatusChangeCallback;
+        if (throwable == null)
+        {
+            log.info("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason);
+        }
+        else
+        {
+            log.info("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason + ", " + throwable.getMessage());
+            throwable.printStackTrace();
+        }
+
+        if (shouldDeviceReconnect(status, statusChangeReason, throwable)) {
+            if (suppliedCallbackPair != null) {
+                suppliedCallbackPair.getKey().execute(DISCONNECTED_RETRYING, NO_NETWORK, throwable, suppliedCallbackPair.getValue());
+            }
+
+            handleRecoverableDisconnection();
+        } else if (suppliedCallbackPair != null) {
+            suppliedCallbackPair.getKey().execute(status, statusChangeReason, throwable, suppliedCallbackPair.getValue());
+        }
+    }
+
+    // This handles the state where the DeviceClient reports that OperationTimeout has expired, and stops retrying; even though the applied RetryPolicy is still valid.
+    protected boolean shouldDeviceReconnect(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable) {
+        return (status == DISCONNECTED && statusChangeReason == RETRY_EXPIRED && throwable instanceof DeviceOperationTimeoutException);
+    }
+
+    public void handleRecoverableDisconnection() {
+        synchronized (lock) {
+            if (connectionStatus == ConnectionStatus.CONNECTED) {
+                new Thread(new Runnable() {
+                    @Override
+                    public void run() {
+                        log.debug("Attempting reconnect for client: " + getClientId() + " ...");
+                        synchronized (lock) {
+                            if (connectionStatus == ConnectionStatus.CONNECTED) {
+                                try {
+                                    closeClient();
+                                } catch (Exception e) {
+                                    log.warn("Client " + getClientId() + " closeNow failed.", e);
+                                } finally {
+                                    connectionStatus = ConnectionStatus.CONNECTING;
```

Comment:
Why wouldn't this be set to DISCONNECTED?

Reply:
The DISCONNECTED status represents a final disconnected (and not going to retry) status. Connecting is basically saying, we are trying to open this connection and it's not yet connected.

Reply:
That might be evident to a sample reader, such as myself, if the enum values had comments describing what they are and when they are used

Reply:
Comments added